### PR TITLE
[Air] Alpa trainer

### PR DESCRIPTION
**Pull request by JiahaoYao**
## Why are these changes needed?

- [x] adding the alpa trainer
- [x] adding mnist example of alpa trainer
- [x] adding the build-kite for alpa from this pr #27142
- [x] adding the test example for alpa trainer 

-----------

### Version I: 
- [x] based on the scaling configuration predefined, we adding `node_resource: 1e-3` to the `ScalingConfig`
- [x] due to the resources prefixed at the beginning, the alpa-trainer will not switch other resources in the ray cluster, even though there might be idle resources. This `features` might not good for the integration of ray tune (say, training multiple small models with lots of tuning). 

### Version II: 
- [x] adding the placement group support in the alpa project: details --- see https://github.com/alpa-projects/alpa/pull/634
- [x] since the support of placement group, currently integration of ray tune is much more flexible to jump into different nodes/resources within ray cluster. 
- [x] Remove the `node_ip: 1e-3` in the Version I implementation
- [x] Reduce a large amount of unnecessary code in the version I implementation.

## Expected results: 

Example of using `AlpaTrainer` on the ray cluster with 4 node (1 gpu for each node):

```python
== Status ==
Current time: 2022-08-04 05:31:48 (running for 00:05:57.09)
Memory usage on this node: 6.5/15.3 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/4 GPUs, 0.0/42.4 GiB heap, 0.0/17.82 GiB objects (0.0/4.0 accelerator_type:T4)
Result logdir: /home/ubuntu/ray_results/AlpaTrainer_2022-08-04_05-25-50
Number of trials: 8/8 (8 TERMINATED)
+-------------------------+------------+-----------------+------------------------+------------------------+--------+------------------+--------------+------------------+
| Trial name              | status     | loc             |   train_loop_config... |   train_loop_config... |   iter |   total time (s) |   train_loss |   train_accuracy |
|-------------------------+------------+-----------------+------------------------+------------------------+--------+------------------+--------------+------------------|
| AlpaTrainer_e6f0c_00000 | TERMINATED | 10.0.2.91:21331 |            0.000209893 |                    200 |    200 |          75.3037 |   0.655804   |         0.848417 |
| AlpaTrainer_e6f0c_00001 | TERMINATED | 10.0.2.91:21416 |            0.000967741 |                    100 |    100 |          52.63   |   0.39745    |         0.8922   |
| AlpaTrainer_e6f0c_00002 | TERMINATED | 10.0.2.91:23928 |            0.00892127  |                    100 |    100 |          53.4502 |   0.150951   |         0.957517 |
| AlpaTrainer_e6f0c_00003 | TERMINATED | 10.0.2.91:1885  |            0.00015414  |                    200 |    200 |          81.0892 |   0.858757   |         0.820817 |
| AlpaTrainer_e6f0c_00004 | TERMINATED | 10.0.2.91:25253 |            0.0465124   |                    200 |    200 |          96.8119 |   0.00558419 |         0.999767 |
| AlpaTrainer_e6f0c_00005 | TERMINATED | 10.0.2.91:20749 |            0.0885211   |                    200 |    200 |          84.1451 |   0.0013098  |         1        |
| AlpaTrainer_e6f0c_00006 | TERMINATED | 10.0.2.91:26332 |            0.0102295   |                    300 |    300 |         127.386  |   0.0442992  |         0.988883 |
| AlpaTrainer_e6f0c_00007 | TERMINATED | 10.0.2.91:10551 |            0.0108772   |                    200 |    200 |          92.89   |   0.0681368  |         0.981617 |
+-------------------------+------------+-----------------+------------------------+------------------------+--------+------------------+--------------+------------------+

```

Version II experiments: (**Difference**: able to jump to different nodes, no longer prefixed!)

```python

== Status ==
Current time: 2022-08-04 22:04:28 (running for 00:01:51.16)
Memory usage on this node: 4.5/15.3 GiB
Using FIFO scheduling algorithm.
Resources requested: 6.0/16 CPUs, 4.0/4 GPUs, 0.0/42.4 GiB heap, 0.0/17.82 GiB objects (0.0/4.0 accelerator_type:T4)
Result logdir: /home/ubuntu/ray_results/AlpaTrainer_2022-08-04_22-02-35
Number of trials: 8/8 (4 PENDING, 2 RUNNING, 2 TERMINATED)
+-------------------------+------------+-----------------+------------------------+------------------------+--------+------------------+--------------+------------------+
| Trial name              | status     | loc             |   train_loop_config... |   train_loop_config... |   iter |   total time (s) |   train_loss |   train_accuracy |
|-------------------------+------------+-----------------+------------------------+------------------------+--------+------------------+--------------+------------------|
| AlpaTrainer_25eeb_00002 | RUNNING    | 10.0.2.46:28918 |            0.00838581  |                    300 |    206 |          64.6613 |    0.0864652 |         0.976117 |
| AlpaTrainer_25eeb_00003 | RUNNING    | 10.0.2.96:5038  |            0.000968097 |                    300 |    160 |          50.1725 |    0.328459  |         0.907683 |
| AlpaTrainer_25eeb_00004 | PENDING    |                 |            0.000225938 |                    200 |        |                  |              |                  |
| AlpaTrainer_25eeb_00005 | PENDING    |                 |            0.000350031 |                    200 |        |                  |              |                  |
| AlpaTrainer_25eeb_00006 | PENDING    |                 |            0.000842312 |                    200 |        |                  |              |                  |
| AlpaTrainer_25eeb_00007 | PENDING    |                 |            0.00279368  |                    300 |        |                  |              |                  |
| AlpaTrainer_25eeb_00000 | TERMINATED | 10.0.2.208:7468 |            0.000158479 |                    100 |    100 |          37.6386 |    1.54061   |         0.737533 |
| AlpaTrainer_25eeb_00001 | TERMINATED | 10.0.2.96:22478 |            0.0522803   |                    100 |    100 |          43.3173 |    0.0234336 |         0.994883 |
+-------------------------+------------+-----------------+------------------------+------------------------+--------+------------------+--------------+------------------+


(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 160, train_loss: 0.3285, train_accuracy: 90.77, epoch_time: 0.212
(AlpaTrainer pid=28918) epoch: 207, train_loss: 0.0860, train_accuracy: 97.62, epoch_time: 0.222
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 161, train_loss: 0.3277, train_accuracy: 90.78, epoch_time: 0.229
Result for AlpaTrainer_25eeb_00002:
  date: 2022-08-04_22-04-28
  done: false
  experiment_id: 3a9054c39a604c43be9a659ecfebb32b
  hostname: ip-10-0-2-46
  iterations_since_restore: 208
  node_ip: 10.0.2.46
  pid: 28918
  time_since_restore: 65.14073848724365
  time_this_iter_s: 0.254016637802124
  time_total_s: 65.14073848724365
  timestamp: 1659650668
  timesteps_since_restore: 0
  train_accuracy: 0.9763166308403015
  train_loss: 0.08563300967216492
  training_iteration: 208
  trial_id: 25eeb_00002
  warmup_time: 0.0036191940307617188

(AlpaTrainer pid=28918) epoch: 208, train_loss: 0.0856, train_accuracy: 97.63, epoch_time: 0.251
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 162, train_loss: 0.3270, train_accuracy: 90.80, epoch_time: 0.219
(AlpaTrainer pid=28918) epoch: 209, train_loss: 0.0852, train_accuracy: 97.64, epoch_time: 0.214
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 163, train_loss: 0.3262, train_accuracy: 90.83, epoch_time: 0.217
(AlpaTrainer pid=28918) epoch: 210, train_loss: 0.0848, train_accuracy: 97.65, epoch_time: 0.234
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 164, train_loss: 0.3255, train_accuracy: 90.85, epoch_time: 0.223
(AlpaTrainer pid=28918) epoch: 211, train_loss: 0.0844, train_accuracy: 97.66, epoch_time: 0.228
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 165, train_loss: 0.3248, train_accuracy: 90.86, epoch_time: 0.226
(AlpaTrainer pid=28918) epoch: 212, train_loss: 0.0840, train_accuracy: 97.67, epoch_time: 0.246
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 166, train_loss: 0.3241, train_accuracy: 90.88, epoch_time: 0.225
(AlpaTrainer pid=28918) epoch: 213, train_loss: 0.0836, train_accuracy: 97.69, epoch_time: 0.253
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 167, train_loss: 0.3233, train_accuracy: 90.90, epoch_time: 0.225
(AlpaTrainer pid=28918) epoch: 214, train_loss: 0.0832, train_accuracy: 97.70, epoch_time: 0.236
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 168, train_loss: 0.3226, train_accuracy: 90.91, epoch_time: 0.230
(AlpaTrainer pid=28918) epoch: 215, train_loss: 0.0828, train_accuracy: 97.71, epoch_time: 0.252
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 169, train_loss: 0.3220, train_accuracy: 90.93, epoch_time: 0.231
(AlpaTrainer pid=28918) epoch: 216, train_loss: 0.0824, train_accuracy: 97.71, epoch_time: 0.239
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 170, train_loss: 0.3213, train_accuracy: 90.95, epoch_time: 0.222
(AlpaTrainer pid=28918) epoch: 217, train_loss: 0.0820, train_accuracy: 97.73, epoch_time: 0.239
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 171, train_loss: 0.3206, train_accuracy: 90.96, epoch_time: 0.208
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 172, train_loss: 0.3199, train_accuracy: 90.99, epoch_time: 0.216
(AlpaTrainer pid=28918) epoch: 218, train_loss: 0.0816, train_accuracy: 97.74, epoch_time: 0.243
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 173, train_loss: 0.3192, train_accuracy: 91.00, epoch_time: 0.215
(AlpaTrainer pid=28918) epoch: 219, train_loss: 0.0813, train_accuracy: 97.75, epoch_time: 0.250
(AlpaTrainer pid=5038, ip=10.0.2.96) epoch: 174, train_loss: 0.3186, train_accuracy: 91.04, epoch_time: 0.218
^C(AlpaTrainer pid=28918) epoch: 220, train_loss: 0.0809, train_accuracy: 97.76, epoch_time: 0.238
``` 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/alpa/alpa_trainer.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        super(AlpaTrainer, self).__init__(
+            scaling_config=scaling_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
+
+    def training_loop(self) -> None:
+        self._train_loop(self._train_loop_config)
```
Comment (amogkam) [inline python/ray/train/alpa/alpa_trainer.py:132]: `self._train_loop(self._train_loop_config, self._datasets, self._resume_from_checkpoint)`

**Comment by JiahaoYao** [comment]
^ pass my local test

**Comment by JiahaoYao** [comment]
enabled:
1.

    trainer = AlpaTrainer(
        train_loop_per_worker=train_func,
        scaling_config=ScalingConfig(num_workers=1, use_gpu=True, resources_per_worker={'CPU': 1, 'GPU': 1}),
    )
    
    tuner = Tuner(
        trainer,
        param_space={
            "train_loop_config": {
                "learning_rate": tune.loguniform(1e-4, 1e-1),
                "batch_size": 5000,
                "momentum": 0.9,
                "num_epochs": tune.choice([1000, 2000, 3000]),
            },
        },
        tune_config=TuneConfig(num_samples=8),
    )
one node multiple alpa process training ( 4 concurrently tuning and the other 4 pending )


2.
```
scaling_config=ScalingConfig(num_workers=2, use_gpu=True, resources_per_worker={'CPU': 1, 'GPU': 2}),
> num_samples=2
```
two node (2 gpu each) training (2 tune process concurrently)

**Comment by JiahaoYao** [comment]
Version II: 
- [x] adding the placement group support in the alpa project: details --- see https://github.com/alpa-projects/alpa/pull/634
- [x] since the support of placement group, currently integration of ray tune is much more flexible to jump into different nodes/resources within ray cluster.

**Comment by JiahaoYao** [comment]
^ this commit enables the version II implementation to work.

**Comment by JiahaoYao** [comment]
^ revert the above commit if the pr in alpa is merged.

**Comment by JiahaoYao** [comment]
the pr for the placement group is merged now, and the monkey patch is removed now!

**Comment by JiahaoYao** [comment]
probably outsource the `AlpaManager` to `alpa.init()`: https://github.com/alpa-projects/alpa/pull/657

**Pull request closed without merging**